### ISSUE TITLE
Feature Request: XGBoost with binary:logistic and other objectives

### ISSUE THREAD

**Issue by metanoid** [alan-turing-institute/MLJModels.jl#96]
The current implementation of the XGBoost wrappers have restrictions on which `objective`s can be set.  
See [here](https://github.com/alan-turing-institute/MLJModels.jl/blob/b76c85726de5d96b847025d9dc92681188e58141/src/XGBoost.jl#L149) and [here](https://github.com/alan-turing-institute/MLJModels.jl/blob/b76c85726de5d96b847025d9dc92681188e58141/src/XGBoost.jl#L376) and [here](https://github.com/alan-turing-institute/MLJModels.jl/blob/b76c85726de5d96b847025d9dc92681188e58141/src/XGBoost.jl#L591)

It would be great to extend the set of supported `objectives` to all possible objectives allowed in [`XGBoost`](https://xgboost.readthedocs.io/en/latest/parameter.html#)

**Comment by ysimillides** [comment]
It was originally like that but was changed based on https://github.com/alan-turing-institute/MLJ.jl/issues/65 . It might be possible to revert the commit / modify it .

**Comment by tlienart** [comment]
Sounds good, @metanoid or @ysimillides could you point out to where the objectives are set in XGBoost.jl ? I can see `binary:logistic` and the poisson one from quickly scraping the repo but  it'd be useful to have pointers so that I can just implement it; @ysimillides could you find the commit? 

I seem to remember we had issues with XGB (a while  back) so maybe that  was the reasoning. But I can try bringing stuff back and see. thanks

**Comment by metanoid** [comment]
I'm no expert, but as far as I can see, the `objective` is always set by passing in a `param` (as `Array{Pair{String,Any},1}`) e.g.

```julia
nfold = 5
param = ["max_depth" => 2,
         "eta" => 1,
         "objective" => "binary:logistic"]
metrics = ["auc"]
nfold_cv(train_X, num_round, nfold, label = train_Y, param = param, metrics = metrics)
```

This array is then consumed by the `XGBoosterSetParam` function [here](https://github.com/dmlc/XGBoost.jl/blob/46b7a20c5d93028d89686205e82aee6f549053c7/src/xgboost_wrapper_h.jl#L17)1 which in turn calls the `xboost` macro [here](https://github.com/dmlc/XGBoost.jl/blob/46b7a20c5d93028d89686205e82aee6f549053c7/src/xgboost_wrapper_h.jl#L10) to turn the parameter pairs into C calls.

**Comment by ablaom** [comment]
Before implementing arbitrary objectives some thought should be given to their relevance to to the scientific type of the target. XGBoost just mashes all these things together, which on the one hand provides flexibility, but on the other hand allows abuse. So, for example, if you are going to use the Poisson objective, then you ought to be using XGBoostCount, not XGBooostRegressor. 

Perhaps @metanoid could say more about his specific needs, including the type of target he is wanting to regresss/classifiy and specific objective functions  he is missing, so we can make these a priority.

**Comment by metanoid** [comment]
Thanks @ablaom.

I'm trying to predict the probability that a given interaction will be successful. In classical statistics this would be a logistic regression problem, since what I want is the probability of success.  So I conceptualised it as having a `Continuous` target, and therefore wanted to try an `XGBoostRegressor` with a `binary:logistic` objective.  

However, at the moment, I'm using `XGBoostClassifier` on a `Multiclass` target (with levels `0.0` and `1.0` representing cases where the event failed or succeeded respectively), and that seems to be working.  I can then use `pdf.(predictions, 1.0)` afterwards to get my required probability of success.  So please don't adjust the roadmap on my account.

This works for me because in my training data, for each observation I am 100% sure of whether the event succeeded or failed (i.e. every observation's target is either exactly 0.0 or 1.0), but there are conceivably situations where this is not true (e.g. for some observations the event has not yet fully completed successfully, but is some real-valued proportion of the way to success), and for these cases it would be nice to be able to use a logistic target for a Regressor.  Unless there's some other way of handling this that I'm missing?

**Comment by ablaom** [comment]
> However, at the moment, I'm using XGBoostClassifier on a Multiclass target (with levels 0.0 and 1.0 representing cases where the event failed or succeeded respectively), and that seems to be working. I can then use pdf.(predictions, 1.0) afterwards to get my required probability of success. So please don't adjust the roadmap on my account.

Yes! That's exactly what you should do. The implementation of XGBoostClassifier just does logistic regression.

> This works for me because in my training data, for each observation I am 100% sure of whether the event succeeded or failed (i.e. every observation's target is either exactly 0.0 or 1.0), but there are conceivably situations where this is not true (e.g. for some observations the event has not yet fully completed successfully, but is some real-valued proportion of the way to success), and for these cases it would be nice to be able to use a logistic target for a Regressor. Unless there's some other way of handling this that I'm missing?

A ha, I understand your use case now. Yes, I agree, we would conceptualize this a `Continuous` target now and a `Deterministic` prediction. But you would like a sigmoid applied to the ensemble prediction, and for training to minimize the training loss function *when applied to the transformed target*, yes? 

Sounds like your proposal to allow XGBoostRegressor access to other objective functions is indeed the best solution.

**Comment by ablaom** [comment]
@ysimillides I do not recommend reverting to an old commit, as there is no way of guaranteeing new bugs are not introduced, and the change is pretty straightforward:

Replace the definition of `clean(::XGBoostRegressor)` with the following:

```julia
function MLJBase.clean!(model::XGBoostRegressor)
    warning = ""
    if model.objective == "count:poisson"
        warning *= "Your `objective` suggests prediction of a "*
        "`Count` variable.\n You may want to consider XGBoostCount instead. "
    elseif model.objective in ["reg:logistic", "binary:logistic",
                               "binary:logitraw", "binary:hinge",
                               "multi:softmax", "multi:softprob"]
        warning *="Your `objective` suggests prediction of a "*
        "`Finite` variable.\n You may want to consider XGBoostClassifier "*
        "instead. "
    end
    return warning
end
```

And a corresponding adjustment to the doc string.

Let's keep the restrictions on XGBoostClassifier and XGBoostCount, as I think they still make sense for now.
